src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] objects;
+    private int size;
+
+    public ArrayList() {
+        objects = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index");
+        }
+        if (size >= objects.length) {
+            grow();
+        }
+        if (index != size) {
+            System.arraycopy(objects, index, objects, index + 1, size - index);
+        }
+        objects[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForAdd(index);
+        return objects[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForAdd(index);
+        objects[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForAdd(index);
+        T value = objects[index];
+        System.arraycopy(objects, index + 1, objects, index, size - index - 1);
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index >= 0) {
+            return remove(index);
+        } else {
+            throw new NoSuchElementException("Element not found");
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        if (objects.length == size) {
+            int newCapacity = size + (size >> 1);
```

Comment:
use of binary operations reduces code readability. let's use constants instead
```suggestion
            int newCapacity = size * RESIZE_FACTOR;
```